modelscope_agent/agent.py

```diff
@@ -1,377 +1,182 @@
-            knowledge_list,
-            append_files=append_files,
-        )
-        function_list = self.prompt_generator.get_function_list(tool_list)
-
-        llm_result, exec_result = '', ''
-
-        idx = 0
-
-        while True:
-            idx += 1
-            llm_artifacts = self.prompt_generator.generate(
-                llm_result, exec_result)
-            if print_info:
-                logger.info(
-                    uuid=kwargs.get('uuid', 'default_user'),
-                    message=f'LLM inputs in round {idx}',
-                    content={'llm_artifacts': llm_artifacts})
-
-            llm_result = ''
-            try:
-                for s in self.llm.stream_generate(llm_artifacts,
-                                                  function_list):
-                    llm_result += s
-                    yield {'llm_text': s}
-            except NotImplementedError:
-                s = self.llm.generate(llm_artifacts)
-                llm_result += s
-                yield {'llm_text': s}
-            except Exception as e:
-                yield {'llm_text': str(e)}
-
-            # parse and get tool name and arguments
-            try:
-                action, action_args = self.action_parser.parse_response(
-                    llm_result)
-            except ValueError as e:
-                yield {'exec_result': f'{e}'}
-                return
-
-            if action is None:
-                # in chat mode, the final result of last instructions should be updated to prompt history
-                _ = self.prompt_generator.generate(llm_result, '')
-                yield {'is_final': True}
-                return
-
-            if action in self.available_tool_list:
-                # yield observation to as end of action input symbol asap
-                yield {'llm_text': 'Observation: '}
-                action_args = self.parse_action_args(action_args)
-                tool = self.tool_list[action]
-
-                # TODO @wenmeng.zwm remove this hack logic for image generation
-                if action == 'image_gen' and self.seed:
-                    action_args['seed'] = self.seed
-                try:
-                    exec_result = tool(**action_args, remote=remote)
-                    yield {'exec_result': exec_result}
-
-                    # parse exec result and update state
-                    self.parse_exec_result(exec_result)
-                except Exception as e:
-                    exec_result = f'Action call error: {action}: {action_args}. \n Error message: {e}'
-                    yield {'exec_result': exec_result}
-                    self.prompt_generator.reset()
-                    return
-            else:
-                exec_result = f"Unknown action: '{action}'. "
-                yield {'exec_result': exec_result}
-                self.prompt_generator.reset()
-                return
-
-    def reset(self):
+    def _detect_tool_by_special_token(self, text: str):
```

Comment:
a better name may be  `extract_tool_info`

Reply:
Deleted